x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformPersistentTasksExecutor.java

```diff
@@ -94,20 +97,67 @@ public TransformPersistentTasksExecutor(
     @Override
     public PersistentTasksCustomMetadata.Assignment getAssignment(TransformTaskParams params, ClusterState clusterState) {
         if (TransformMetadata.getTransformMetadata(clusterState).isResetMode()) {
-            return new PersistentTasksCustomMetadata.Assignment(null,
-                "Transform task will not be assigned as a feature reset is in progress.");
+            return new PersistentTasksCustomMetadata.Assignment(
+                null,
+                "Transform task will not be assigned as a feature reset is in progress."
+            );
         }
         List<String> unavailableIndices = verifyIndicesPrimaryShardsAreActive(clusterState, resolver);
         if (unavailableIndices.size() != 0) {
+            // check if the latest index got created
+            if (unavailableIndices.contains(TransformInternalIndexConstants.LATEST_INDEX_NAME)) {
+                CountDownLatch indexCreationLatch = new CountDownLatch(1);
+                ActionListener<Void> templateCheckListener = ActionListener.wrap(
+                    aVoid -> {
+                        logger.debug(
+                            "Successfully created transform internal index: [{}]",
+                            TransformInternalIndexConstants.LATEST_INDEX_NAME
+                        );
+                        indexCreationLatch.countDown();
+                    },
+                    error -> {
+                        Throwable cause = ExceptionsHelper.unwrapCause(error);
+                        String reason = "Not starting transform ["
+                            + params.getId()
+                            + "], because the transform internal index could not be created";
+
+                        if (internalIndexCreationFailure == null) {
+                            // only log it once
+                            logger.error(reason, cause);
+                        }
+                        internalIndexCreationFailure = reason + ", cause: " + cause.getMessage();
+                        indexCreationLatch.countDown();
+                    }
+                );
+                TransformInternalIndex.ensureLatestIndexAndTemplateInstalled(clusterService, client, templateCheckListener);
+
+                // note: wait for the index to be created, but do not block the assigner too long, it will be called again
+                // e.g. by the periodic re-check every 30s
+                try {
+                    indexCreationLatch.await(WAIT_FOR_INDEX_CREATION_TIMEOUT_SEC, TimeUnit.SECONDS);
```

Comment:
We can't do this here. 
First, this blocks EVERY persistent task assignment method.

Second, this is done on a cluster state change listener callback. Which means it blocks handling cluster state changes.